### PR TITLE
chop plan generation output off and leave only summary when publishing

### PR DESCRIPTION
**Pull request by alex-goncharov**
This trims the `terraform plan` state refresh output and keeps only the plan summary for publishing. Also shoots some warnings with task agent to be seen on the main page/tab of the build summary (super handy, trust me)

    - add decorator that keeps terraform command output as is
    - expose task logging in task agent

Supersedes #248

### CODE REVIEW

**Comment by alex-goncharov** [comment]
@charleszipp this is a stab at parsing the plan, which completely relies on the plan  output and does not run display. On the other hand it wants to always run with -detailed-exit code. But since change detection was always there and it does not work completely without the exit code, we should be fine.

File: tasks/terraform-cli/src/commands/tf-plan.ts [hunk 1]
```diff
@@ -27,41 +43,88 @@ export class TerraformPlan implements ICommand {
...
+    private getPlanHasChangesExitCode(commandOptions: string | undefined): number {
         return this.hasDetailedExitCode(commandOptions) ? 2 : 0;
     }
 
-    private setPlanHasChangesVariable(ctx: ITaskContext, exitCode: number): void{
-        let planHasChanges:boolean = exitCode === this.getPlanHasChangesExitCode(ctx.commandOptions);
-        ctx.setVariable("TERRAFORM_PLAN_HAS_CHANGES", planHasChanges.toString(), false);
+    private setPlanHasChangesVariable(ctx: ITaskContext, exitCode: number): void {
+        let planHasChanges: boolean = exitCode === this.getPlanHasChangesExitCode(ctx.commandOptions);
+        ctx.setVariable("TERRAFORM_PLAN_HAS_CHANGES", (exitCode === terraformPlanOkHasChanges).toString(), false);
```
Comment (charleszipp) [inline tasks/terraform-cli/src/commands/tf-plan.ts:128]: Lets make sure this still reports as it did before if `publishPlanResults` does not exist (see earlier comment about only adding detailed exit code when publishPlanResults exists). Currently, if they dont provide -detailed-exitcode and terraform returns exit code 0, its set to true. There should be existing tests that check this behavior that should pass if this works.

File: tasks/terraform-cli/src/commands/tf-plan.ts [hunk 2]
```diff
@@ -27,41 +43,88 @@ export class TerraformPlan implements ICommand {
...
     }
 
-    private getPlanSuccessfulExitCodes(commandOptions: string | undefined): number[]{
-        let successCodes: number[] = [];
-        successCodes.push(0);
-        if(this.hasDetailedExitCode(commandOptions))
-            successCodes.push(2);
-        return successCodes;
+    private planSummaryReport(toAdd: string, toUpdate: string, toDestroy: string) {
+        this.taskAgent.warning(`This plan is going to create ${toAdd} resources.`)
```
Comment (charleszipp) [inline tasks/terraform-cli/src/commands/tf-plan.ts:77]: Inject ILogger via the constructor. See [tf-output for example](https://github.com/charleszipp/azure-pipelines-tasks-terraform/blob/f4af431726ed3d3f795ac89ad3dbd38963f1d031/tasks/terraform-cli/src/commands/tf-output.ts#L10). You can then do `this.logger.warning`. That will do what you added to task-agent.

You will also need to update task.ts to make sure the TerraformPlan command is instantiated with the logger instance from task.ts.

**Review by charleszipp — CHANGES_REQUESTED** [review]
Thanks @alex-goncharov for reworking the approach. I like the approach and think this will open some interesting possibilities. 

Only significant issue is with the way detailed exitcode is being added. If we add this always, it will change how the environment variables are set. This could break pipelines if they have conditions that only check for == or != 0. Lets add this on their behalf only if publishPlanResults exists so they can opt-into the breaking change for these env vars.

File: tasks/terraform-cli/src/commands/tf-plan.ts (hunk 1 above)
Comment (charleszipp) [reply tasks/terraform-cli/src/commands/tf-plan.ts:128]: The existing tests I mentioned are failing now but, should start passing once this behaves as it did before when `publishPlanResults` is not set.

Comment (alex-goncharov) [reply tasks/terraform-cli/src/commands/tf-plan.ts]: fixed

File: tasks/terraform-cli/src/commands/tf-plan.ts [hunk 3]
```diff
@@ -27,41 +43,88 @@ export class TerraformPlan implements ICommand {
     async exec(ctx: ITaskContext): Promise<CommandResponse> {
         const provider = this.getProvider(ctx);
         const successCodes = this.getPlanSuccessfulExitCodes(ctx.commandOptions)
-        const options = await new RunWithTerraform(ctx)            
+        const options = await new RunWithTerraform(ctx)
             .withProvider(ctx, provider)
             .withSecureVarFile(this.taskAgent, ctx.secureVarsFileId, ctx.secureVarsFileName)
             .withCommandOptions(ctx.commandOptions)
+            .withDetailedExitCode(ctx.commandOptions)
```
[the remark this answers is not in the document]
Comment (alex-goncharov) [reply tasks/terraform-cli/src/commands/tf-plan.ts]: reverted logic back for the case when publishing is not configured and no -detailed-exitcode is given. The detailed code is forcefully set when publishing is requested though.

File: tasks/terraform-cli/src/commands/tf-plan.ts (hunk 2 above)
Comment (alex-goncharov) [reply tasks/terraform-cli/src/commands/tf-plan.ts]: fixed

**Review by charleszipp — APPROVED** [review]
LGTM! Thanks so much @alex-goncharov for all your hard work on this. I think this will be very helpful to others that dont want to look at 100's or 1000's of lines of refresh logs in the plan view.

**Pull request merged**